cloud/services/virtualmachines/virtualmachines.go

```diff
@@ -158,6 +159,21 @@ func (s *Service) Reconcile(ctx context.Context, spec interface{}) error {
 		virtualMachine.Identity = &compute.VirtualMachineIdentity{
 			Type: compute.ResourceIdentityTypeSystemAssigned,
 		}
+	} else if vmSpec.Identity == infrav1.VMIdentityUserAssigned {
+		if len(vmSpec.UserAssignedIdentities) == 0 {
+			return errors.Wrapf(err, "cannot create VM: The user-assigned identity provider ids must not be null or empty for 'UserAssigned' identity type.")
+		}
+		// UserAssignedIdentities - The list of user identities associated with the Virtual Machine.
+		// The user identity dictionary key references will be ARM resource ids in the form:
+		// '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ManagedIdentity/userAssignedIdentities/{identityName}'.
+		userIdentitiesMap := make(map[string]*compute.VirtualMachineIdentityUserAssignedIdentitiesValue)
```

Comment:
I think you can allocate with an initial size based on the length of vmSpec.UserAssignedIdentities?

Reply:
yes, makes sense